libpmem2/index.md

```diff
@@ -0,0 +1,209 @@
+or create an annonymous map.
+
+The next key step in this example is to set the granularity using
+`pmem2_config_set_required_store_granularity`.
+Granularity must be one of the three values: `PMEM2_GRANULARITY_BYTE`,
+`PMEM2_GRANULARITY_CACHE_LINE`, `PMEM2_GRANULARITY_PAGE`.
+In this case, we set a maximum permitted granularity to `PMEM2_GRANULARITY_PAGE`.
+In practice, it means that the program will flush data using e.g. CLFLUSH or
+will use eADR feature when available on the platform, but if is not available,
+then `msync()` is also accepted and no error will occur.
+
+More detailed informations about granularity concept and each option
+can be found  <a href="../manpages/linux/master/libpmem2/{{ page.title }}.7.html">here.</a>
+
+The last step just create a mapping using described above
+config and source. Underlying function of `pmem2_map()` call `mmap(2)`
+on POSIX or `CreateFileMapping()` on Windows to memory map the entire file
+concerning the permissible granularity.
+
+When the mapping is created we are ready to write into it.
+
+<code data-gist-id='wlemkows/c7dc06875ee4aa9857020eba01114e75' data-gist-file='basic.c' data-gist-line='60-63' data-gist-highlight-line='63' data-gist-hide-footer='true'></code>
+
+Using the getters like `pmem2_map_get_size`, `pmem2_map_get_addres`
+we can easy read information about created mapping.
+
+The novel thing about pmem is you can copy to it directly, like any
+memory.  The `strcpy()` call shown on line 63 above is just the usual
+_libc_ function that stores a string to memory. If this example program
+were to be interrupted either during or just after the `strcpy()` call,
+you can't be sure which parts of the string made it all the way to the
+media. It might be none of the string, all of the string, or somewhere
+in-between. Also, there's no guarantee the string will make it to
+the media in the order it was stored! For longer ranges, it is just as
+likely that portions copied later make it to the media before earlier
+portions. (So don't write code like the example above and then expect
+to check for zeros to see how much of the string was written.)
+
+How can a string get stored in a seemingly random order?
+The reason is that until a flush function like `msync()`
+has returned successfully, the normal cache pressure that happens
+on an active system can push changes out to the media at any time
+in any order. Most processors have barrier instructions (like
+`SFENCE` on the Intel platform) but those instructions deal with
+ordering in the visibility of stores to other threads, not with
+the order that changes reach persistence.
+The only barrier for flushing to persistence is function returned by
+`pmem2_get_persist_fn()` as shown below.
+
+<code data-gist-id='wlemkows/c7dc06875ee4aa9857020eba01114e75' data-gist-file='basic.c' data-gist-line='65-66' data-gist-highlight-line='65' data-gist-hide-footer='true'></code>
+
+The libpmem2 function `pmem2_get_persist_fn` automatically determine
```

Comment:
automatically decides what's the most appropriate mechanism for flushing data onto the underlying storage. This means that if the mapping does not support user-space flushing, the persist function will fall back to using the OS primitives for synchronizing data.

Something like that.